fx-src/FirefoxStorage.js

```diff
@@ -0,0 +1,484 @@
+      this._dbconn = null;
+      this.statements = {};
+    }
+  }
+}
+
+export default class FirefoxAdapter extends BaseAdapter {
+  constructor(dbname) {
+    super();
+    this.operations = [];
+    this.busy = false;
+
+
+    // attempt creation - we don't need to use a transaction here since our
+    // operation queue means these will happen before other operations.
+    const statements = ["CREATE TABLE IF NOT EXISTS collection_metadata (collection_name TEXT PRIMARY KEY, last_modified INTEGER) WITHOUT ROWID;",
+          "CREATE TABLE IF NOT EXISTS collection_data (collection_name TEXT, record_id TEXT, record TEXT);",
+          "CREATE UNIQUE INDEX IF NOT EXISTS unique_collection_record ON collection_data(collection_name, record_id);"];
+
+    this.kintoStorage = new KintoStorage(dbname);
+
+    for (let stmt of statements) {
+      this.executeUpdate(stmt);
+    }
+  }
+
+  executeUpdate(sql) {
+    debug("requesting to execute statement: "+sql);
+    this.executeOperation(function(kintoStorage, complete) {
+      debug("executing statement: "+sql);
+      const statement = kintoStorage.getStatement(sql);
+
+      statement.executeAsync({
+        handleError: function(aError) {
+          debug("Error: " + aError.message);
+        },
+
+        handleCompletion: function(aReason) {
+          if (aReason != Ci.mozIStorageStatementCallback.REASON_FINISHED) {
+            debug("Query canceled or aborted!");
+          } else {
+            debug("Query complete");
+          }
+          complete();
+        }
+      });
+    });
+  }
+
+  executeOperation(op) {
+    debug("executeOperation");
+    if (!this.busy) {
+      debug("Executing operation now");
+      this.busy = true;
+      function executeNextOperation() {
+        debug("executeNextOperation");
+        const next = this.operations.shift();
+        if (next) {
+          debug("executing queued operation");
+          next(this.kintoStorage, executeNextOperation.bind(this));
+        } else {
+          debug("cleaning up");
+          this.busy = false;
+          this.kintoStorage.requestAsyncClose();
+          debug("work queue complete");
+        }
+      }
+      op(this.kintoStorage, executeNextOperation.bind(this));
+    } else {
+      debug("queuing operation");
+      this.operations.push(op);
+    }
+  }
+
+  clear() {
+    return new Promise((resolve, reject) => {
+      this.executeOperation((kintoStorage, complete) => {
+        debug("kinto::clear");
+        // clear all of the data for this adapter
+        const statement = kintoStorage.getStatement("DELETE FROM collection_data WHERE collection_name = :collection_name;");
+        statement.params.collection_name = kintoStorage.getName();
+
+        // execute the statement
+        statement.executeAsync({
+          handleResult: function(aResultSet) {
+            debug("A result set was not expected");
+          },
+
+          handleError: function(aError) {
```

Comment:
This all gets much less painful when you use Sqlite.jsm, which returns promises.